### PR TITLE
AVM1: Some Activation cleanup

### PR DESCRIPTION
**Pull request by moulins**
- Remove unused `GcCell`s on `Activation` and `Scope`;
- Remove the `globals` parameter from `Activation::from_nothing`
- Reduce allocations in `Activation::from_nothing`:
  - Pre-allocate the global scope on the `Avm1` context so that it can be reused;
  - Reuse the global constant pool instead of allocating a new empty one; this doesn't matter as no bytecode is directly executed on a `from_nothing` activation without bringing its own constant pool.

------

`Activation::from_nothing` now does a single allocation (in `Scope::new_local_scope`), but I don't know if this can be removed.

Naïvely, an activation created with this method shouldn't be used to directly execute bytecode or to define locals, and so it should be fine to reuse directly the global scope (and indeed, the tests still pass after doing this change), but I'm not sure this reasoning is actually correct.

### CODE REVIEW

**Comment by moulins** [comment]
After a careful look at all callers of `Activation::from_nothing` and `Activation::from_stub`, I've concluded that none of them are directly defining new local variables on the returned `Activation`, so removing the extra local scope is fine, at the price of restricting what can be done with these `Activation`s.

As such, this PR now makes `Activation::from_nothing` fully allocation-free.

File: core/src/avm1/activation.rs [hunk 1]
```diff
@@ -2834,7 +2809,6 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
             self.set_variable(name, value)
         } else {
             let scope = self.scope;
```
Comment (relrelb) [inline core/src/avm1/activation.rs:2811]: nit: Let's line this up to `self.scope.define_local(...)`

File: core/src/avm1/scope.rs [hunk 2]
```diff
@@ -30,10 +29,10 @@ pub enum ScopeClass {
 }
 
 /// Represents a scope chain for an AVM1 activation.
-#[derive(Debug, Collect)]
+#[derive(Copy, Clone, Debug, Collect)]
```
Comment (relrelb) [inline core/src/avm1/scope.rs:32]: I prefer keeping large structs non-`Copy`, since cloning them isn't cheap.

File: core/src/avm1/scope.rs [hunk 3]
```diff
@@ -60,67 +59,31 @@ impl<'gc> Scope<'gc> {
...
-            let grandparent = parent.read().parent;
-            if let Some(grandparent) = grandparent {
-                parent = grandparent;
-            } else {
-                break;
-            }
+        if scope.class == ScopeClass::Target {
+            scope.values = clip;
+        } else {
+            scope.parent = scope.parent.map(|p| Self::new_target_scope(p, clip, mc));
```
Comment (relrelb) [inline core/src/avm1/scope.rs:71]: I think recursion can be avoided here using a simple loop.

File: core/src/avm1/scope.rs (hunk 2 above)
Comment (kmeisthax) [reply core/src/avm1/scope.rs:32]: How large is 'large', here? We have two GC pointers - one optional, and one enumclass'd; and a binary enum. I doubt it exceeds 32 bytes.

File: core/src/avm1/activation.rs (hunk 1 above)
Comment (moulins) [reply core/src/avm1/activation.rs:2811]: I did this at the start, but I ended up with weird borrowck errors; I could try changing it to `self.scope()...` though.

File: core/src/avm1/scope.rs (hunk 3 above)
Comment (moulins) [reply core/src/avm1/scope.rs:71]: Unfortunately no :(
With a loop you'd need to allocate the scope chain and *then* replace the target (which is what the old code did), but now that `Scope` is immutable there's no way to replace the scope object once allocated (because it's behind a `Gc`).

File: core/src/avm1/scope.rs (hunk 2 above)
Comment (moulins) [reply core/src/avm1/scope.rs:32]: With alignment, it's 24 bytes on x86-64, and 12 bytes on WASM, so definitely not a "big" struct.

File: core/src/avm1/activation.rs (hunk 1 above)
Comment (moulins) [reply core/src/avm1/activation.rs]: Fixed!

File: core/src/avm1/scope.rs (hunk 2 above)
Comment (relrelb) [reply core/src/avm1/scope.rs:32]: Actually I consider structs larger than a pointer size as "big" enough. In my opinion implicit clones for big structs hurt code readability (See `new_target_scope` for example, it's very unclear that `scope` is a clone of `parent`, at least for me).

File: core/src/avm1/scope.rs [hunk 4]
```diff
@@ -30,10 +29,10 @@ pub enum ScopeClass {
...
 
 /// Represents a scope chain for an AVM1 activation.
-#[derive(Debug, Collect)]
+#[derive(Copy, Clone, Debug, Collect)]
 #[collect(no_drop)]
 pub struct Scope<'gc> {
-    parent: Option<GcCell<'gc, Scope<'gc>>>,
+    parent: Option<Gc<'gc, Scope<'gc>>>,
     class: ScopeClass,
     values: Object<'gc>,
```
Comment (relrelb) [inline core/src/avm1/scope.rs:37]: Might not be related, but I think `values` should be reduced to just `PropertyMap<'gc, Value<'gc>>`. `Object<'gc>` includes many functionality that doesn't make sense in the context of `Scope` (e.g. prototype, watchers, interfaces etc.). Though this might belong to a separate PR.

File: core/src/avm1/scope.rs [hunk 5]
```diff
@@ -60,67 +59,31 @@ impl<'gc> Scope<'gc> {
...
     /// scope has been replaced with another given object.
     pub fn new_target_scope(
-        mut parent: GcCell<'gc, Self>,
+        parent: Gc<'gc, Self>,
         clip: Object<'gc>,
         mc: MutationContext<'gc, '_>,
-    ) -> GcCell<'gc, Self> {
-        let mut bottom_scope = None;
-        let mut top_scope: Option<GcCell<'gc, Self>> = None;
+    ) -> Gc<'gc, Self> {
```
Comment (relrelb) [inline core/src/avm1/scope.rs:65]: `new_target_scope` can actually return `Self`, then the `mc` parameter is redundant.

File: core/src/avm1/scope.rs [hunk 6]
```diff
@@ -49,7 +48,7 @@ impl<'gc> Scope<'gc> {
     }
 
     /// Construct a child scope of another scope.
-    pub fn new_local_scope(parent: GcCell<'gc, Self>, mc: MutationContext<'gc, '_>) -> Scope<'gc> {
+    pub fn new_local_scope(parent: Gc<'gc, Self>, mc: MutationContext<'gc, '_>) -> Scope<'gc> {
```
Comment (relrelb) [inline core/src/avm1/scope.rs:51]: Let's use `Self` instead of `Scope<'gc>`.

File: core/src/avm1/scope.rs (hunk 2 above)
Comment (kmeisthax) [reply core/src/avm1/scope.rs:32]: Gotcha. So the rule is `size_of::<T>() > size_of::<usize>()` should not be `Copy`.

I wonder if this could be a custom lint...

Comment (moulins) [reply core/src/avm1/scope.rs:32]: Well, by this rule `&[T]` shouldn't be `Copy` :D 
Will remove the Copy impl anyways, I don't think it's actually used much.

File: core/src/avm1/scope.rs (hunk 4 above)
Comment (moulins) [reply core/src/avm1/scope.rs:37]: What about `with` blocks? Don't they trigger watchers and all that jazz?

File: core/src/avm1/scope.rs (hunk 5 above)
Comment (moulins) [reply core/src/avm1/scope.rs:65]: `new_target_scope` needs the `mc` parameters to allocate the parent scopes. Will change for `new_with_scope` though.

File: core/src/avm1/scope.rs (hunk 4 above)
Comment (relrelb) [reply core/src/avm1/scope.rs:37]: Since this isn't really related to this PR, let's defer the discussion to another time (which I don't have currently).

**Review by relrelb — APPROVED** [review]
@moulins Let's merge this. Ping me once conflicts are resolved.

**Comment by moulins** [comment]
Rebased the PR on #8605.

**Review by kmeisthax — APPROVED** [review]
Will merge after the soundpeaks PR finishes going through CI
